src/charmed_kubeflow_chisme/testing/cos_integration.py

```diff
@@ -109,22 +102,30 @@ async def deploy_and_assert_grafana_agent(
     await model.wait_for_idle(apps=[GRAFANA_AGENT_APP], status="blocked", timeout=5 * 60)
 
 
-async def _check_metrics_endpoint(app: Application, metrics_endpoint: str) -> None:
-    """Check metrics endpoint accessibility.
+def _check_url(url: str, port: int, path: str) -> bool:
+    """Check port and path in url."""
+    output = urlparse(url)
+    return output.port == port and output.path == path
 
-    Checking accessibility of metrics endpoint from grafana-agent-k8s. If metrics endpoint is
-    defined as `*:5000/metrics` it will be changed to `<app-name>.<namespace>.svc:5000/metrics`.
+
+async def _check_metrics_endpoint(app: Application, port: int, path: str) -> None:
+    """Check metrics targets accessibility.
+
+    Checking if metrics port and metrics path are dedine in any targes of Grafana agent.
+    E.g. curl localhost:12345/agent/api/v1/metrics/targets return with success status and it's
+    configured as expected.
     """
-    if metrics_endpoint.startswith("*"):
-        url = f"http://{app.name}.{app.model.name}.svc{metrics_endpoint[1:]}"
-    else:
-        url = f"http://{metrics_endpoint}"
+    cmd = f"curl -m 5 -sS {GRAFANA_AGENT_API_TARGETS}"
+    grafana_agent_unit = app.model.applications[GRAFANA_AGENT_APP].units[0]
+    log.debug("testing metrics endpoint with cmd: `%s`", cmd)
+    output = await _run_on_unit(grafana_agent_unit, cmd)
+    targets = yaml.safe_load(output.results["stdout"])
+    log.info("metrics targets definened at %s:\n%s", grafana_agent_unit.name, targets)
 
-    cmd = f"curl -m 5 -sS {url}"
-    grafana_agent_app = app.model.applications[GRAFANA_AGENT_APP]
-    log.info("testing metrics endpoint with cmd: `%s`", cmd)
-    for unit in grafana_agent_app.units:
-        await _run_on_unit(unit, cmd)
+    assert targets["status"] == "success"
+    assert any(
+        _check_url(data["endpoint"], port, path) for data in targets["data"]
+    ), f":{port}{path} was not found in any Grafana agent targets"
```

Comment:
I think we are also interested in is in knowing if those targets correspond to the charm under test AND if their state (or health) is "up".

We could also make some assertions to the metrics endpoint to check if the expected metrics endpoint is the one that grafana agent actually receives.

Reply:
> I think we are also interested in is in knowing if those targets correspond

Test are designed in way that user can chcek if any targets is defined, so it's uo to charm developer to test all targets relevat to the charm.

> AND if their state (or health) is "up"

Do you think that overall status is not enough? I know that it can be missleading, if grafana agent exporter or any other failed. Maybe I can remove check 'targets["status"] == "success"' and chcek only if specific target is up. WDYT?

> We could also make some assertions to the metrics endpoint to check if the expected metrics endpoint is the one that grafana agent actually receives.

I do not get this part. Can you explainwhat do you mean?

Reply:
I changed the behavior to get target data from grafana-agent API for specific application and then validate that those are with `state == "up"` and contain proper endpoint and labels.

Reply:
> I changed the behavior to get target data from grafana-agent API for specific application and then validate that those are with `state == "up"` and contain proper endpoint and labels.

Sorry, maybe I got too late to this comment, but yes, this is what we want in the end because the "status" may show success, but the state of each target is what matters to us as this will indicate that the metrics endpoint is reachable, so that ensures that any change we make in a PR regarding the metrics endpoint will be tested and confirmed.